### PR TITLE
search: paginated repository resolution

### PR DESCRIPTION
**Pull request by tsenart**
This PR is the last of series to introduce paginated repo resolution in search.

- We remove `backend.ListSearchable` and its use in repository resolution because global searches on Sourcegraph.com no longer depend on it. This in turn simplifies the code in preparation for cursor based pagination.
- In order to be able to remove the blocking `resolveRepositories` in `doResults`, repository search had to be migrated to the new `Job` interface. This is implemented by `run.RepoSearch`.
- We migrate all existing `Job` implementations to use `searchrepos.Resolver.Paginate`.
- Alerts for missing repo revs and no resolved repos errors are now handled by `searchResolver.errorToAlert`.

Part of #26995

### CODE REVIEW

**Comment by tsenart** [comment]
Tomorrow morning I'd like to write some thorough tests of `Paginate`, fix some existing brittle tests and handle any feedback you leave meanwhile. This PR is targeting @rvantonder's open PR which migrates symbol search to the Job interface.

**Comment by tsenart** [comment]
Best reviewed by hiding whitespace.

**Review by rvantonder — COMMENTED** [review]
Had a glance and in principle looks good! Very exciting. Looks like backend tests fail (nice job doing this with a `backend-integration` branch) so could you rebase and rerequest review when its green?

File: internal/search/repos/repos.go [hunk 1]
```diff
@@ -82,98 +139,111 @@ func (r *Resolver) Resolve(ctx context.Context, op search.RepoOptions) (Resolved
...
-			NoArchived:             op.NoArchived,
-			OnlyArchived:           op.OnlyArchived,
-			NoPrivate:              op.Visibility == query.Public,
-			OnlyPrivate:            op.Visibility == query.Private,
-			SearchContextID:        searchContext.ID,
-			UserID:                 searchContext.NamespaceUserID,
-			OrgID:                  searchContext.NamespaceOrgID,
-			IncludeUserPublicRepos: searchContext.ID == 0 && searchContext.NamespaceUserID != 0,
-		}
+	if len(repos) == 0 && len(op.Cursors) == 0 {
```
Comment (tsenart) [inline internal/search/repos/repos.go:178]:
```suggestion
	if len(repos) == 0 && len(op.Cursors) == 0 { // Is the first page empty?
```

File: internal/search/repos/repos.go [hunk 2]
```diff
@@ -82,98 +136,111 @@ func (r *Resolver) Resolve(ctx context.Context, op search.RepoOptions) (Resolved
...
+		return Resolved{}, ErrNoResolvedRepos
+	}
+
+	var next types.Cursors
+	if len(repos) == limit+1 {
```
Comment (tsenart) [inline internal/search/repos/repos.go:183]:
```suggestion
	if len(repos) == limit+1 { // Do we have a next page?
```

**Review by stefanhengl — APPROVED** [review]
Reviewed synchronously together with @tsenart. LGTM!

**Comment by tsenart** [comment]
@rvantonder: PTAL again

File: cmd/frontend/graphqlbackend/search_alert.go [hunk 3]
```diff
@@ -238,130 +232,6 @@ func (e *errOverRepoLimit) Error() string {
 	return "Too many matching repositories"
 }
 
-func (r *searchResolver) errorForOverRepoLimit(ctx context.Context) *errOverRepoLimit {
```
Comment (tsenart) [inline cmd/frontend/graphqlbackend/search_alert.go:241]: This wasn't used anywhere anymore, even before this PR.

File: internal/search/run/run.go [hunk 4]
```diff
@@ -29,7 +30,7 @@ type SearchInputs struct {
 // may not be required, depending on the job. E.g., a global search job does not
 // require upfront repository resolution).
 type Job interface {
-	Run(context.Context, streaming.Sender, []*search.RepositoryRevisions) error
+	Run(context.Context, streaming.Sender, searchrepos.Pager) error
```
Comment (rvantonder) [inline internal/search/run/run.go:33]: sweet.

File: internal/search/run/repository.go [hunk 5]
```diff
@@ -121,6 +132,24 @@ func SearchRepositories(ctx context.Context, args *search.TextParameters, limit
...
+
+	return repos.Paginate(ctx, func(page *searchrepos.Resolved) error {
+		return searchRepositories(ctx, s.Args, page.RepoRevs, s.Limit, stream)
+	})
+}
+
+func (*RepoSearch) Name() string {
+	return "Repo"
+}
+
```
Comment (rvantonder) [inline internal/search/run/repository.go:152]: Nice, I was thinking to do this for the repo search (basically just bundle all of `search.TextParameters` for now and then get rid of the cruft we don't need. I'm just a teeeny bit reluctant to throw this change in with the pagination (it would be more conservative to first convert the job and monitor Sourcegraph.com) but since backend integration tests pass we have _reasonable_ confidence this is good. I'm just mentioning this because during migrating things, _I_ once messed up passing the stream around without limit even though I was extra extra careful, and the issue only manifested when Sourcegraph.com OOMs :-)

File: cmd/frontend/graphqlbackend/search_results.go [hunk 6]
```diff
@@ -1763,6 +1731,8 @@ func (r *searchResolver) doResults(ctx context.Context, args *search.TextParamet
 			return waitGroup(args.ResultTypes.Without(result.TypeCommit) == 0)
 		case "Symbol":
 			return waitGroup(args.ResultTypes.Without(result.TypeSymbol) == 0)
+		case "Repo":
+			return waitGroup(args.ResultTypes.Without(result.TypeRepo) == 0)
```
Comment (rvantonder) [inline cmd/frontend/graphqlbackend/search_results.go:1735]: ⚠️ This should be `waitGroup(true)` to preserve behavior of previous code. ⚠️ 

Camden and I have been thinking about the waitgroup logic and it can probably go away, but that should be something we deal with in a different PR. It is/was used as a state that says whether a job's result is optional or not before returning results, but with streaming this matters less and maybe totally unneeded.

**Review by rvantonder — APPROVED** [review]
LGTM. There's one important thing to fix up with `waithGroup(true)` but not going to block approval for that.

File: internal/search/commit/commit_new.go [hunk 7]
```diff
@@ -347,8 +351,10 @@ type DiffCommitError struct {
 	Max        int
 }
 
-type RepoLimitError DiffCommitError
-type TimeLimitError DiffCommitError
+type (
+	RepoLimitError DiffCommitError
+	TimeLimitError DiffCommitError
+)
```
[the remark this answers is not in the document]
Comment (tsenart) [reply internal/search/commit/commit_new.go:357]: Yep, it was, but I think the only one, so I felt lazy to take it out :)

File: cmd/frontend/graphqlbackend/search_results_test.go [hunk 8]
```diff
@@ -248,9 +213,6 @@ func TestSearchResults(t *testing.T) {
 		if !calledReposListMinimalRepos {
 			t.Error("!calledReposListMinimalRepos")
 		}
-		if !calledSearchRepositories {
-			t.Error("!calledSearchRepositories")
-		}
```
[the remark this answers is not in the document]
Comment (tsenart) [reply cmd/frontend/graphqlbackend/search_results_test.go:253]: 🧇

File: cmd/frontend/graphqlbackend/search.go [hunk 9]
```diff
@@ -279,6 +278,10 @@ func (r *searchResolver) resolveRepositories(ctx context.Context, options search
 		tr.Finish()
 	}()
 
+	// TODO(tsenart)
+	// 	  Move per-request caching to searchrepos.Pager, and make the cache key the RepoOptions struct
+	//    (including the Cursor field). Once that is done, we can get rid of this `resolveRepositories`
+	//    method which only adds caching around searchrepos.Resolver.
```
Comment (keegancsmith) [inline cmd/frontend/graphqlbackend/search.go:284]: Do we even still need the caching? I think that complicated caching logic was added for the suggestions graphqlbackend code. I'm not sure if that is still relevant?

File: cmd/frontend/graphqlbackend/search_alert.go [hunk 10]
```diff
@@ -238,130 +232,6 @@ func (e *errOverRepoLimit) Error() string {
...
-				proposedQueries = append(proposedQueries, &searchQueryDescription{
-					description: fmt.Sprintf("in the repository %s", strings.TrimPrefix(pathToPropose, "github.com/")),
-					query:       newExpr,
-					patternType: r.PatternType,
-				})
-			}
-		}
-	}
-	return buildErr(proposedQueries, description)
-}
```
Comment (keegancsmith) [inline cmd/frontend/graphqlbackend/search_alert.go:363]: TODO check if we lose this logic

File: cmd/frontend/graphqlbackend/search_alert.go [hunk 11]
```diff
@@ -482,85 +322,87 @@ func capFirst(s string) string {
 	}, s)
 }
 
-func alertForError(err error) *searchAlert {
+func (r *searchResolver) errorToAlert(ctx context.Context, err error) (*searchAlert, error) {
```
Comment (keegancsmith) [inline cmd/frontend/graphqlbackend/search_alert.go:325]: nit: it isn't clear what is now being used from searchResolver. Making this part of searchResolver is kinda a step backwards, since we want to move things like this out of graphqlbackend (whoops looks like we havent yet for some weird reason?). What is actually being used in r? Is it just the query, can we explicitly pass that in?

File: internal/search/repos/repos.go [hunk 12]
```diff
@@ -38,20 +40,72 @@ type Resolved struct {
 	RepoSet         map[api.RepoID]types.MinimalRepo
 	MissingRepoRevs []*search.RepositoryRevisions
 	OverLimit       bool
+
+	// Next points to the next page of resolved repository revisions. It will
+	// be nil if there are no more pages left.
+	Next types.Cursors
```
Comment (keegancsmith) [inline internal/search/repos/repos.go:46]: nit: likely unrelated, but why is this `Cursors` plural? Even if your cursor is multi-dimensional it still kinda feels like a single cursor from the clients point of view.

**Review by keegancsmith — COMMENTED** [review]
I'm mid review but gotta go for lunch. Gonna post it so far in case the in flight review is useful before I finish.

File: cmd/frontend/graphqlbackend/search.go (hunk 9 above)
Comment (tsenart) [reply cmd/frontend/graphqlbackend/search.go:284]: Nope, I don't think we do, specially since we'll remove the suggestions code before the next release. I'll change the wording in this TODO to reflect that.

File: internal/search/repos/repos.go (hunk 12 above)
Comment (tsenart) [reply internal/search/repos/repos.go:46]: I will rename this type to `MultiCursor` for clarity.

File: cmd/frontend/graphqlbackend/search_alert.go (hunk 11 above)
Comment (tsenart) [reply cmd/frontend/graphqlbackend/search_alert.go:325]: It's not just the query unfortunately. @stefanhengl had the same feedback initially, but addressing it would result in an even larger change, since the `alertForNoResolvedRepos` is defined on `searchResolver` and needs a few things, like a db handle and search inputs.

File: cmd/frontend/graphqlbackend/search_alert.go (hunk 10 above)
Comment (tsenart) [reply cmd/frontend/graphqlbackend/search_alert.go:363]: This was unused so I removed it. https://github.com/sourcegraph/sourcegraph/pull/27093/files/062181e0faa1aad5880a354bf1b3ffbf57809c80#r743964387

File: internal/database/repos.go [hunk 13]
```diff
@@ -1519,31 +1522,47 @@ func parsePattern(p string) ([]*sqlf.Query, error) {
...
-	case string(RepoListCreatedAt):
-		conds = append(conds, sqlf.Sprintf("created_at "+direction+" %s", c.Value))
-	default:
-		return nil, errors.Errorf("missing or invalid cursor: %q %q", c.Column, c.Value)
+	if len(columns) == 0 {
+		return nil, nil
 	}
-	return conds, nil
+
+	return sqlf.Sprintf(fmt.Sprintf("(%s) %s (%%s)", strings.Join(columns, ", "), operator), sqlf.Join(values, ", ")), nil
```
Comment (keegancsmith) [inline internal/database/repos.go:1565]: I think this should work if you change `columns` to `[]*sqlf.Query` and operator to `*sqlf.Query`:

```suggestion
	return sqlf.Sprintf("(%s) %s (%s)", sqlf.Join(columns, ", "), operator, sqlf.Join(values, ", ")), nil
```

File: internal/search/repos/repos.go [hunk 14]
```diff
@@ -82,98 +136,111 @@ func (r *Resolver) Resolve(ctx context.Context, op search.RepoOptions) (Resolved
 		return Resolved{}, err
 	}
 
-	var searchableRepos []types.MinimalRepo
+	tr.LazyPrintf("Repos.List - start")
```
Comment (keegancsmith) [inline internal/search/repos/repos.go:139]: nit between this printf and the next all you do is create the listoptions struct, so either git rid of this one or the next since they won't provide any useful timing info

File: internal/search/repos/repos.go [hunk 15]
```diff
@@ -82,98 +136,111 @@ func (r *Resolver) Resolve(ctx context.Context, op search.RepoOptions) (Resolved
...
-			NoArchived:             op.NoArchived,
-			OnlyArchived:           op.OnlyArchived,
-			NoPrivate:              op.Visibility == query.Public,
-			OnlyPrivate:            op.Visibility == query.Private,
-			SearchContextID:        searchContext.ID,
-			UserID:                 searchContext.NamespaceUserID,
-			OrgID:                  searchContext.NamespaceOrgID,
-			IncludeUserPublicRepos: searchContext.ID == 0 && searchContext.NamespaceUserID != 0,
-		}
+	if len(repos) == 0 && len(op.Cursors) == 0 { // Is the first page empty?
```
Comment (keegancsmith) [inline internal/search/repos/repos.go:178]: somewhere else it was documented that op.Cursors is nil meaning last page. Maybe those docs need to be updated to mean empty or nil?

File: internal/search/repos/repos.go [hunk 16]
```diff
@@ -82,98 +136,111 @@ func (r *Resolver) Resolve(ctx context.Context, op search.RepoOptions) (Resolved
...
+				c.Value = strconv.FormatInt(int64(last.Stars), 10)
+			case "id":
+				c.Value = strconv.FormatInt(int64(last.ID), 10)
 			}
-		}
 
-		repos, err = r.DB.Repos().ListMinimalRepos(ctx, options)
-		tr.LazyPrintf("Repos.List - done")
+			if o.Descending {
+				c.Direction = "prev"
```
Comment (keegancsmith) [inline internal/search/repos/repos.go:196]: nit: is direction recent? It being a plain string seems somewhat error prone. Previously I saw we had a Descending boolean, what made you switch away from boolean?

File: internal/search/run/repository.go [hunk 17]
```diff
@@ -121,6 +132,24 @@ func SearchRepositories(ctx context.Context, args *search.TextParameters, limit
...
+	Limit int
+}
+
+func (s *RepoSearch) Run(ctx context.Context, stream streaming.Sender, repos searchrepos.Pager) error {
+	ctx, stream, cleanup := streaming.WithLimit(ctx, stream, s.Limit)
+	defer cleanup()
+
+	return repos.Paginate(ctx, func(page *searchrepos.Resolved) error {
+		return searchRepositories(ctx, s.Args, page.RepoRevs, s.Limit, stream)
+	})
```
Comment (keegancsmith) [inline internal/search/run/repository.go:146]: @tsenart: just pointing out this is quite beautiful, nice solution! Previously we batch fetched and then cached that copy of the repos list. This was really hard to untangle. Love that we now instead just respeak to the DB for each call site that needs repos. I believe right now that should be really small, except for some deprecated code paths. Nice!

**Review by keegancsmith — APPROVED** [review]
So excited to finally change our relationship with resolving repos. This has been a tough nut to crack! LGTM with the high level caveats pointed out below:

### Repository Counts

We report the number of repositories we search over in the UI. This PR will now have that count change depending on how far along we got in pagination before we hit result limits. An easy way to reproduce this once landed is to do a global search for a common term. I suspect each time you refresh the page you will get a different number. Numbers being consistent an important UX principle we adhere to so users have confidence in our search. So this will be a UX regression we need to follow-up on. Ideas

- Calculate count like we do for exclude. Possible to combine that into a single query more than likely with the use of group by.
- Stop showing the number in the UX. Personally I find the number useful to gauge order of magnitude I am searching over, so I think this will need more thought if we go with it.

### More DB interaction on global searches

Right now we use a cached copy of the global repos when doing global searches on sourcegraph.com. I believe this PR changes this behaviour to speaking to the DB until we load all 2m repos or the search is done. To me this poses the most likely risk for reverting this PR. It won't make searches slower directly, but I imagine will increase DB load.

File: internal/search/repos/repos.go (hunk 16 above)
Comment (tsenart) [reply internal/search/repos/repos.go:196]: Nope, the Cursor type already existed as separate fields in database.RepoListOptions, I just promoted it to its own struct in #26990 so that they could be passed around. I agree, but would like to avoid making this change in this PR.

File: internal/search/run/repository.go (hunk 17 above)
Comment (tsenart) [reply internal/search/run/repository.go:146]: Glad you like it!

File: internal/search/repos/repos.go (hunk 15 above)
Comment (tsenart) [reply internal/search/repos/repos.go:178]: No, `Resolved.Next` being nil means it's the last page. `RepoOptions.Cursors` being nil means we're requesting the first page.

File: internal/search/repos/repos.go (hunk 14 above)
Comment (tsenart) [reply internal/search/repos/repos.go:139]: Good catch, will remove.

File: internal/database/repos.go (hunk 13 above)
Comment (tsenart) [reply internal/database/repos.go:1565]: That would make the columns use up bind parameters which isn't needed and makes debugging the query harder.

**Comment by tsenart** [comment]
> Right now we use a cached copy of the global repos when doing global searches on sourcegraph.com. I believe this PR changes this behaviour to speaking to the DB until we load all 2m repos or the search is done. To me this poses the most likely risk for reverting this PR. It won't make searches slower directly, but I imagine will increase DB load.

No, we don't load up all 2M repos. If we did, the stats would be correct, but we don't. We stop as soon as we have enough search results. The only user of repo resolution in global searches is now `type:repo` search, which we do by default.

File: internal/database/repos.go (hunk 13 above)
Comment (keegancsmith) [reply internal/database/repos.go:1565]: you can do what you do here where the format string is the value. So

```
columns = append(columns, sqlf.Sprintf(column))
```
